File: web/src/types/Layers.ts
interface Layer {
  type: string;
}

export interface ConvLayer extends Layer {
  type: "conv";
  filters: number;
  kernel_size: number;
  strides: number;
  padding: string;
  activation: string;
}

export interface PoolLayer extends Layer {
  type: "pool";
  pool_size: number;
  strides: number;
  padding: string;
}

export interface Dropout extends Layer {
  type: "dropout";
  rate: number;
}

export interface BatchNorm extends Layer {
  type: "batchnorm";
}

export interface Flatten extends Layer {
  type: "flatten";
}

export interface Dense extends Layer {
  type: "dense";
  units: number;
  activation: string;
}

export interface TextVectorization extends Layer {
  type: "textvectorization";
  max_tokens: number;
  output_sequence_length: number;
}

export interface Embedding extends Layer {
  type: "embedding";
  input_dim: number;
  output_dim: number;
}

export interface GlobalAveragePooling1D extends Layer {
  type: "globalaveragepooling1d";
}

File: modal-function/type_defs/layers.py
from pydantic import BaseModel
from typing import List, Union, Literal

class Layer(BaseModel):
    type: str

class ConvLayer(Layer):
    type: Literal["conv"] = "conv"
    filters: int
    kernel_size: int
    strides: int
    padding: str
    activation: str

class PoolLayer(Layer):
    type: Literal["pool"] = "pool"
    pool_size: int
    strides: int
    padding: str

class Dropout(Layer):
    type: Literal["dropout"] = "dropout"
    rate: float

class BatchNorm(Layer):
    type: Literal["batchnorm"] = "batchnorm"

class Flatten(Layer):
    type: Literal["flatten"] = "flatten"

class Dense(Layer):
    type: Literal["dense"] = "dense"
    units: int
    activation: str

class TextVectorization(Layer):
    type: Literal["textvectorization"] = "textvectorization"
    max_tokens: int
    output_sequence_length: int

class Embedding(Layer):
    type: Literal["embedding"] = "embedding"
    input_dim: int
    output_dim: int

class GlobalAveragePooling1D(Layer):
    type: Literal["globalaveragepooling1d"] = "globalaveragepooling1d"

